### PR TITLE
[GHS] Introduce General HTML Support

### LINKED ISSUE #9856: [GHS] Stabilize and release a production-ready General HTML support feature
## 📝 Provide a description of the new feature

Placeholder for all followups for the GHS feature that we'll need to close to make it production-ready.

List of issues:

*   ~\[GHS\] Applying attributes to existing features - table~ - #9914 :white\_check\_mark:
*   ~\[GHS\] Applying attributes to existing features - image~ - #9916
*   \[GHS\] Applying attributes to existing features - lists - #9917
*   ~\[GHS\] Applying attributes to existing features - media embed~ - #9918
*   \[GHS\] Support for empty elements - #9960
*   \[GHS\] Handling content being a part of the object - #9936
*   ~\[GHS\] Add a demo with all HTML features enabled~ - #9987

---

If you'd like to see this feature implemented, add a 👍 reaction to this post.

### LINKED ISSUE #9970: [GHS] General HTML Support feature
## 📝 Provide a description of the new feature

The General HTML Support ("GHS") feature will allow developers to easily enable HTML features that are not explicitly supported by any other dedicated CKEditor 5 plugins.

Some examples of HTML features that should be easily enabled using General HTML Support include:

* The `<section>`, `<article>`, and `<div>` elements.
* The `<audio>`, `<video>`, and `<iframe>` elements.
* The `<span>` and `<cite>` elements.
* Some of the attributes on existing dedicated CKEditor 5 features:
	* `data-*` and `id` attributes on e.g. `<p>` and `<h1-h6>`,
	* `style` and `classes` on e.g. `<strong>` and `<a>`.

The enabled HTML features could be loaded (e.g. via `editor.setData()`), pasted, output (e.g. via `editor.getData()`), and will be visible in the editing area. Such content could also be edited in the editor, although, to a limited extend.

---

If you'd like to see this feature implemented, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by jacekbogdanski**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Feature (html-support): Introduced General HTML Support feature. Closes #9970.

---

### Additional information

Let's see if we can improve anything before the release.

Since most code has been already reviewed, I suppose it doesn't make much sense to do the full review again, but it would be good to check a couple of things to make sure we didn't miss anything during development:

* manual tests
* guide
* API docs
* overall code quality, rebase artifacts, outdated comments, etc

### CODE REVIEW

**Comment by jacekbogdanski** [comment]
Let's also cover https://github.com/ckeditor/ckeditor5/issues/9951 within this PR.

File: packages/ckeditor5-html-support/src/conversionutils.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+	for ( const key in newValue ) {
+		// Merge classes.
+		if ( Array.isArray( newValue[ key ] ) ) {
+			result[ key ] = Array.from( new Set( [ ...oldValue[ key ], ...newValue[ key ] ] ) );
+		}
+
+		// Merge attributes or styles.
+		else {
+			result[ key ] = { ...oldValue[ key ], ...newValue[ key ] };
```
Comment (maxbarnas) [inline packages/ckeditor5-html-support/src/conversionutils.js:53]: The `newValue` / `oldValue` names don't seem to make sense here.

```suggestion
* @param {Object} attributesA
* @param {Object} attributesB
* @returns {Object}
*/
export function mergeViewElementAttributes( attributesA, attributesB ) {
	const result = cloneDeep( attributesA );

	for ( const key in attributesB ) {
		// Merge classes.
		if ( Array.isArray( attributesB[ key ] ) ) {
			result[ key ] = Array.from( new Set( [ ...attributesA[ key ], ...attributesB[ key ] ] ) );
		}

		// Merge attributes or styles.
		else {
			result[ key ] = { ...attributesA[ key ], ...attributesB[ key ] };
```

File: packages/ckeditor5-html-support/src/datafilter.js [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+	/**
+	 * Batch load of the filtering configuration.
+	 *
+	 * @private
+	 * @param {Array.<module:engine/view/matcher~MatcherPattern>} config Filtering configuration.
+	 * @param {Function} handleAttributes Callback handling the way the attributes should be processed.
+	 */
+	_loadConfig( config, handleAttributes ) {
+		for ( const pattern of config ) {
+			// MatcherPattern allows omitting `name` to not narrow searches to specific elements.
```
Comment (maxbarnas) [inline packages/ckeditor5-html-support/src/datafilter.js:202]:
```suggestion
			// MatcherPattern allows omitting `name` to widen the search of elements.
```

File: packages/ckeditor5-html-support/src/datafilter.js [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+		const value = getValue( prop );
+		if ( value !== undefined ) {
+			attributesObject[ prop ] = getValue( prop );
+		}
+	}
+
+	return attributesObject;
+}
+
+// Matcher by default has to match **all** patterns to count it as an actual match. By splitting the pattern
```
Comment (maxbarnas) [inline packages/ckeditor5-html-support/src/datafilter.js:569]:
```suggestion
// Matcher by default has to match **all** patterns to count it as an actual match. Splitting the pattern
```

File: packages/ckeditor5-html-support/src/schemadefinitions.js [hunk 4]
```diff
@@ -0,0 +1,812 @@
...
+		},
+		{
+			model: 'htmlTr',
+			view: 'tr',
+			modelSchema: {
+				allowIn: [ 'htmlTable', 'htmlThead', 'htmlTbody' ],
+				isBlock: true
+			}
+		},
+		// TODO can also include text.
```
Comment (maxbarnas) [inline packages/ckeditor5-html-support/src/schemadefinitions.js:163]: We're keeping the `TODO` comments here, right?

**Review by maxbarnas — COMMENTED** [review]
Looks great!

File: packages/ckeditor5-html-support/src/schemadefinitions.js (hunk 4 above)
Comment (jacekbogdanski) [reply packages/ckeditor5-html-support/src/schemadefinitions.js:163]: I would keep them for now until we cover https://github.com/ckeditor/ckeditor5/pull/9891 so we can review those definitions again and verify how they should behave.

File: packages/ckeditor5-html-support/docs/features/general-html-support.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+For instance, the dedicated {@link features/basic-styles#available-text-styles bold} feature offers a toolbar button used to make the selected text bold. Together with the {@link features/autoformat autoformatting feature}, it also allows for applying bold style to content by typing a Markdown shortcode (`**foo**`) in the editor. The {@link features/headings headings} feature offers a dropdown from which the user can choose a heading level and ensures that pressing <kbd>Enter</kbd> at the end of a heading creates a new paragraph (and not another heading).
+
+The General HTML Support does not offer any UI for the enabled features and takes only the basic semantics of a given feature into account. If you enable support for `<div>`s via GHS, the user will not be able to create `<div>`s from the editor UI. The GHS will know that a `<div>` is a container element, so it can wrap other blocks (like paragraphs) but cannot be used inline (next to e.g. a `<strong>` element). There is no more, however.
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply packages/ckeditor5-html-support/docs/features/general-html-support.md:86]: It was already highlighted in https://github.com/ckeditor/ckeditor5/issues/9951#issuecomment-868421819 so I'm not touching it now as AFAIK @Mgsy is working on it.

File: packages/ckeditor5-html-support/src/conversionutils.js (hunk 1 above)
Comment (jacekbogdanski) [reply packages/ckeditor5-html-support/src/conversionutils.js:53]: Maybe we could go with `target` and `source` names instead? `attributesB` are replacing `attributesA`, so it seems like a common naming convention.

**Comment by jacekbogdanski** [comment]
Docs update looks good to me 👍🏻

**Pull request merged**